File: hello-csharp/src/Server/Protocol.cs
namespace AudioStreamServer;

/// <summary>Command type classification.</summary>
public enum CommandType { Stream, Data, Query }

/// <summary>Stream lifecycle commands.</summary>
public enum StreamCommand { Create, Complete, Close }

/// <summary>Data transfer commands.</summary>
public enum DataCommand { Read }

/// <summary>Query commands.</summary>
public enum QueryCommand { GetStatus, ListStreams }

/// <summary>Parsed command information.</summary>
public record CommandInfo(CommandType CmdType, StreamCommand? StreamCmd, DataCommand? DataCmd, QueryCommand? QueryCmd)
{
    private static readonly Dictionary<string, CommandInfo> Map = new()
    {
        ["CREATE"]       = new(CommandType.Stream, StreamCommand.Create,   null, null),
        ["COMPLETE"]     = new(CommandType.Stream, StreamCommand.Complete, null, null),
        ["CLOSE"]        = new(CommandType.Stream, StreamCommand.Close,    null, null),
        ["READ"]         = new(CommandType.Data,   null, DataCommand.Read, null),
        ["GET_STATUS"]   = new(CommandType.Query,  null, null, QueryCommand.GetStatus),
        ["LIST_STREAMS"] = new(CommandType.Query,  null, null, QueryCommand.ListStreams),
    };

    public static CommandInfo Lookup(string command) =>
        Map.TryGetValue(command, out var info)
            ? info
            : throw new ArgumentException($"Unknown command: {command}");
}

File: hello-kotlin/src/main/kotlin/server/Protocol.kt
package server

/** Command type classification. */
enum class CommandType { STREAM, DATA, QUERY }

/** Stream lifecycle commands. */
enum class StreamCommand { CREATE, COMPLETE, CLOSE }

/** Data transfer commands. */
enum class DataCommand { READ }

/** Query commands. */
enum class QueryCommand { GET_STATUS, LIST_STREAMS }

/** Parsed command information. */
data class CommandInfo(
    val cmdType: CommandType,
    val streamCmd: StreamCommand? = null,
    val dataCmd: DataCommand? = null,
    val queryCmd: QueryCommand? = null,
) {
    companion object {
        private val map = mapOf(
            "CREATE"       to CommandInfo(CommandType.STREAM, streamCmd = StreamCommand.CREATE),
            "COMPLETE"     to CommandInfo(CommandType.STREAM, streamCmd = StreamCommand.COMPLETE),
            "CLOSE"        to CommandInfo(CommandType.STREAM, streamCmd = StreamCommand.CLOSE),
            "READ"         to CommandInfo(CommandType.DATA, dataCmd = DataCommand.READ),
            "GET_STATUS"   to CommandInfo(CommandType.QUERY, queryCmd = QueryCommand.GET_STATUS),
            "LIST_STREAMS" to CommandInfo(CommandType.QUERY, queryCmd = QueryCommand.LIST_STREAMS),
        )

        fun lookup(command: String): CommandInfo =
            map[command] ?: throw IllegalArgumentException("Unknown command: $command")
    }
}
